### PR TITLE
#420: More operations in string theory

### PR DESCRIPTION
**Pull request by kfriedberger**
This PR adds support for conversion between strings and integers based on [code points](https://en.wikipedia.org/wiki/Code_point#In_character_encoding).
See https://github.com/sosy-lab/java-smt/discussions/420 for the requirement of this API extension in String theory.

Open issues from here:
- Princess has some missing features for String theory, aka crashes on simple tests.
- how does SMTLIB escape escaped Unicode? e.g., how could a user specify "\u{1234}" as String directly?

### CODE REVIEW

**Review by daniel-raffler — COMMENTED** [review]
Looks good to me! There is some overlap with https://github.com/sosy-lab/java-smt/issues/412, but I think it's fine to do it as part of this pull request.

And I think we still need to add the new operations to `FunctionDeclarationKind` for the visitor?

File: src/org/sosy_lab/java_smt/basicimpl/AbstractStringFormulaManager.java [hunk 1]
```diff
@@ -299,4 +352,18 @@ public StringFormula toStringFormula(IntegerFormula number) {
   }
 
   protected abstract TFormulaInfo toStringFormula(TFormulaInfo pParam);
+
+  @Override
+  public IntegerFormula toCodePoint(StringFormula number) {
+    return wrapInteger(toCodePoint(extractInfo(number)));
+  }
+
+  protected abstract TFormulaInfo toCodePoint(TFormulaInfo pParam);
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/basicimpl/AbstractStringFormulaManager.java:361]: We only use the suffix '...Impl' for ambiguous method names, typically in conjunction with lists and generics. The current naming should be valid.

File: src/org/sosy_lab/java_smt/basicimpl/AbstractStringFormulaManager.java [hunk 2]
```diff
@@ -299,4 +352,18 @@ public StringFormula toStringFormula(IntegerFormula number) {
...
+
+  @Override
+  public StringFormula fromCodePoint(IntegerFormula number) {
+    return wrapString(fromCodePoint(extractInfo(number)));
+  }
+
+  protected abstract TFormulaInfo fromCodePoint(TFormulaInfo pParam);
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/basicimpl/AbstractStringFormulaManager.java:368]: same as above

File: src/org/sosy_lab/java_smt/basicimpl/AbstractStringFormulaManager.java [hunk 3]
```diff
@@ -40,11 +53,55 @@ protected RegexFormula wrapRegex(TFormulaInfo formulaInfo) {
...
+    checkArgument(
+        isCodePointInRange(value),
+        "String constant is out of supported Unicode range (Plane 0-2).");
     return wrapString(makeStringImpl(value));
   }
 
+  /** returns whether all Unicode characters in Planes 0-2. */
+  static boolean isCodePointInRange(String str) {
+    return str.codePoints().allMatch(codePoint -> 0x00000 <= codePoint && codePoint <= 0x2FFFF);
+  }
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/basicimpl/AbstractStringFormulaManager.java:71]: sure. done.

File: src/org/sosy_lab/java_smt/basicimpl/AbstractStringFormulaManager.java [hunk 4]
```diff
@@ -40,11 +53,55 @@ protected RegexFormula wrapRegex(TFormulaInfo formulaInfo) {
...
+      }
+      int codePoint = Integer.parseInt(hexCodePoint, 16);
+      checkArgument(
+          0x00000 <= codePoint && codePoint <= 0x2FFFF,
+          "SMTLIB does only specify Unicode letters from Planes 0-2");
+      matcher.appendReplacement(sb, Character.toString(codePoint));
+    }
+    matcher.appendTail(sb);
+    return sb.toString();
+  }
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/basicimpl/AbstractStringFormulaManager.java]: Regex just has problems with plain "\\u". :-) We will solve this issue later, as it depends partially on the escaping in solver Z3 or CVC4.

File: src/org/sosy_lab/java_smt/test/SolverVisitorTest.java [hunk 5]
```diff
@@ -744,6 +745,9 @@ public void stringInStringFormulaVisit() throws SolverException, InterruptedExce
             .add(smgr.charAt(x, offset))
             .add(smgr.toStringFormula(offset))
             .add(smgr.concat(x, y, z));
+    if (solverToUse() != Solvers.PRINCESS) { // TODO Princess crashes with MatchError of IFunApp
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/test/SolverVisitorTest.java:748]: reported here: https://github.com/uuverifiers/princess/issues/18

File: src/org/sosy_lab/java_smt/test/ModelEvaluationTest.java [hunk 6]
```diff
@@ -147,9 +147,42 @@ public void testGetBooleansEvaluation() throws SolverException, InterruptedExcep
...
+        Lists.newArrayList(smgr.makeString("hello \u00e6@\u20ac \u1234 \u4321")));
+
+    // TODO Z3 and CVC4 seem to break escaping on invalid Unicode Strings.
+    /*
+      evaluateInModel(
+        smgr.equal(smgr.makeVariable("x"), smgr.makeString("\\u")),
+        smgr.makeVariable("x"),
+        Lists.newArrayList("\\u"),
+        Lists.newArrayList(smgr.makeString("\\u")));
+    */
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/test/ModelEvaluationTest.java:183]: https://github.com/sosy-lab/java-smt/issues/412 seems to be a good place.

File: src/org/sosy_lab/java_smt/api/FunctionDeclarationKind.java [hunk 7]
```diff
@@ -300,6 +300,8 @@ public enum FunctionDeclarationKind {
   STR_IN_RE,
   STR_TO_INT,
   INT_TO_STR,
+  STR_FROM_CODE,
+  STR_TO_CODE,
   STR_LT,
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/api/FunctionDeclarationKind.java:305]: I added the mentioned operations already here.

**Pull request merged**